sui/src/benchmark.rs

```diff
@@ -0,0 +1,289 @@
+            if let Err(e) = server.join().await {
+                error!("Server ended with an error: {e}");
+            }
+        });
+    });
+
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_stack_size(32 * 1024 * 1024)
+        .worker_threads(usize::min(num_cpus::get(), 24))
+        .build()
+        .unwrap();
+
+    // Wait for server start
+    sleep(Duration::from_secs(3));
+
+    // Follower to observe batches
+    let follower_network_client = network_client.clone();
+    thread::spawn(move || {
+        let runtime = Builder::new_multi_thread()
+            .enable_all()
+            .thread_stack_size(32 * 1024 * 1024)
+            .build()
+            .unwrap();
+
+        runtime.block_on(async move { run_follower(follower_network_client).await });
+    });
+
+    sleep(Duration::from_secs(3));
+
+    // Run load
+    let (elapsed, resp) =
+        runtime.block_on(async move { send_tx_chunks(txes, network_client, connections).await });
+
+    let _: Vec<_> = resp
+        .par_iter()
+        .map(|q| check_transaction_response(deserialize_message(&(q.as_ref().unwrap())[..])))
+        .collect();
+    MicroBenchmarkResult::Throughput {
+        chunk_throughput: 1_000_000.0 * num_transactions as f64 / elapsed as f64,
+    }
+}
+
+fn run_latency_microbench(
+    network_client: NetworkClient,
+    network_server: NetworkServer,
+    connections: usize,
+    use_move: bool,
+    committee_size: usize,
+    db_cpus: usize,
+
+    num_chunks: usize,
+    chunk_size: usize,
+    period_us: u64,
+) -> MicroBenchmarkResult {
+    assert!(
+        (num_chunks * chunk_size % connections) == 0,
+        "num_transactions must {} be multiple of number of TCP connections {}",
+        num_chunks * chunk_size,
+        connections
+    );
+    let mut tx_cr = TransactionCreator::new(committee_size, db_cpus);
+
+    // These TXes are to load the network
+    let load_gen_txes = tx_cr.generate_transactions(connections, use_move, chunk_size, num_chunks);
+
+    // These are tracer TXes used for measuring latency
+    let tracer_txes = tx_cr.generate_transactions(1, use_move, 1, num_chunks);
+
+    // Make multi-threaded runtime for the authority
+    thread::spawn(move || {
+        let runtime = Builder::new_multi_thread()
+            .enable_all()
+            .thread_stack_size(32 * 1024 * 1024)
+            .build()
+            .unwrap();
+
+        runtime.block_on(async move {
+            let server = spawn_authority_server(network_server, tx_cr.authority_state).await;
+            if let Err(e) = server.join().await {
+                error!("Server ended with an error: {e}");
+            }
+        });
+    });
+
+    // Wait for server start
+    sleep(Duration::from_secs(3));
+
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_stack_size(32 * 1024 * 1024)
+        .worker_threads(usize::min(num_cpus::get(), 24))
```

Comment:
Depending on where this will be run, we might get more reproducible results if we use, say, half the CPUs.

Reply:
(Also the authority should probably use only half as well, looks like its using all cores right now)

Reply:
I tried with half and still got same level of variance